### PR TITLE
resync missions in edgeclusters periodically

### PR DESCRIPTION
**Pull request by chenqianfzh**
With this PR, the edgecontroller will align the missions in edgeclusters with the missions in the cloud side.

In more details: Every time an edgecluster object is added or updated, edgecontroller will look into the missions deployed in the edge cluster. If there is any discrepancy between the missions of the edge side and the cloud side, the edge controller will send an MissionList message to the ege side. The MetaManager will update the mission records saved in its DB after receiving the MissionList message. Finally, the clusterd will pick up the changes and update the missions in its managed edge cluster.

### Verification
The PR is verified with the following test scenario :
1. Start the cloudcore
2. keep the edgecore off and deploy a new mission A
3. start edgecore
It is verified that the mission A is deployed to the edge cluster

4. stop the edgecore process
5. delete the mission A and deploy a new mission B in the cloud cluster.
6. restart the edgecore process, wait like 20 seconds
7. It is verified that in the edge cluster, mission A is removed and mission B is active.

8. Manually delete the mission B in the edge cluster.
9. wait 20 seconds.
10. It is verified that mission B is re-deployed and still active in the edge cluster.

### Possible improvement
In this PR, the mission list message will be sent out only when it is necessary, namely when the edge controller finds the missions in the edge is different from the missions in the cloud side. For the PoC purpose, I compare the missions of edge and cloud by their names. Therefore, if a mission's spec is manually changed in the edge side with the mission name untouched, the cloud side cannot detect it.

 I am thinking this issue might be solved by clusterd sending a Mission Spec hash value in the edge-cluster status message, where the edgecontroller get the information of missons running on the edge cluster. Then the edge controller can also check whether the mission Spec is different from that in the cloud side. It is left for further discussion .

### CODE REVIEW

**Comment by pdgetrf** [comment]
@jshaofuturewei this is POC code so please ignore the test failures and just focus on ideas being verified.

File: edge/pkg/clusterd/clusterd.go [hunk 1]
```diff
@@ -152,6 +152,10 @@ func (e *clusterd) syncCloud() {
 		}
 		klog.V(4).Infof("result content is %s", result.Content)
 		_, resType, resID, err := util.ParseResourceEdge(result.GetResource(), result.GetOperation())
+		if err != nil {
+			klog.Errorf("Failed in edge resource parsing: %v", err)
```
Comment (pdgetrf) [inline edge/pkg/clusterd/clusterd.go:156]: Failed --> failed

Comment (pdgetrf) [reply edge/pkg/clusterd/clusterd.go:156]: or just stay consistent with the form

File: edge/pkg/metamanager/process.go [hunk 2]
```diff
@@ -230,6 +232,56 @@ func (m *metaManager) processInsert(message model.Message) {
...
+			Type:  constants.ResourceTypeMission,
+			Value: string(data)}
+		err = dao.InsertOrUpdate(meta)
+		if err != nil {
+			klog.Errorf("Update meta failed, %v", meta)
+			continue
+		}
+
+		if _, exists := strayKeys[daoKey]; exists {
+			//dao.DeleteMetaByKey
```
Comment (pdgetrf) [inline edge/pkg/metamanager/process.go:270]: is this comment still needed?

**Review by pdgetrf — APPROVED** [review]
minor comments. LGTM

File: cloud/pkg/edgecontroller/controller/downstream.go [hunk 3]
```diff
@@ -585,14 +589,14 @@ func (dc *DownstreamController) syncMissions() {
...
-				nodeName, ok := key.(string)
+				clusterName, ok := key.(string)
 				if !ok {
 					klog.Warning("Failed to assert key to sting")
 					return true
 				}
 				msg := model.NewMessage("")
 				msg.SetResourceVersion(mission.ResourceVersion)
-				resource, err := messagelayer.BuildResource(nodeName, "default", model.ResourceTypeMission, mission.Name)
+				resource, err := messagelayer.BuildResource(clusterName, "default", model.ResourceTypeMission, mission.Name)
```
Comment (jshaofuturewei) [inline cloud/pkg/edgecontroller/controller/downstream.go:599]: Is "default" the model namespace?

**Review by jshaofuturewei — APPROVED** [review]

File: edge/pkg/metamanager/process.go (hunk 2 above)
Comment (chenqianfzh) [reply edge/pkg/metamanager/process.go:270]: deleted the comment

File: edge/pkg/clusterd/clusterd.go (hunk 1 above)
Comment (chenqianfzh) [reply edge/pkg/clusterd/clusterd.go:156]: updated.

File: cloud/pkg/edgecontroller/controller/downstream.go (hunk 3 above)
Comment (chenqianfzh) [reply cloud/pkg/edgecontroller/controller/downstream.go:599]: Actually, this is a workaround. The work to build a cluster-scope resource, where the namespace should be "", is not done. So the message will not be received if the namespace is "". Yet the message will be correctly received as long as both sides have the same namespace name. So I just put it as "default" in the PoC work.

**Pull request merged**